File: Apps/iOS/Features/Harmonica/HarmonicaViewState.swift
enum HarmonicaViewState: Equatable {
    case preparingSound
    case ready(Playable)
    case soundUnavailable(String)

    // MARK: - Playable

    struct Playable: Equatable {
        let holes: [Hole]
        let key: Key
        let fingerMarks: FingerMarks
        let cup: Cup?
        let demo: Demo
        let shapingPad: ShapingPad
    }

    // MARK: - Hole

    struct Hole: Equatable, Identifiable {
        let id: Int
        let label: String
        let note: String
        let effect: String
        let lit: LitHalf?
    }

    // MARK: - LitHalf

    enum LitHalf: Hashable {
        case top
        case bottom
    }

    // MARK: - Key

    struct Key: Equatable {
        let label: String
        let position: Double
        let highestPosition: Double
    }

    // MARK: - FingerMarks

    struct FingerMarks: Equatable {
        let isDrawnAtTheContactWidth: Bool
        let drawsOnlyTheDecidingFinger: Bool
    }

    // MARK: - Cup

    struct Cup: Equatable {
        let label: String
        let closed: Double
    }

    // MARK: - Demo

    struct Demo: Equatable {
        let label: String
        let tunes: [Tune]
        let isPlaying: Bool
    }

    // MARK: - Tune

    struct Tune: Equatable, Identifiable {
        let id: Int
        let name: String
    }

    // MARK: - ShapingPad

    struct ShapingPad: Equatable {
        let pitchLabel: String
        let vibratoLabel: String
        let isPitchShapingAvailable: Bool
        let placement: ShapingPadPlacement
        let size: ShapingPadSize
    }
}

File: Web/src/features/harmonica/harmonicaViewState.ts
export type HarmonicaViewState =
    | { readonly kind: 'preparingSound' }
    | { readonly kind: 'ready'; readonly playable: PlayableHarmonica }
    | { readonly kind: 'soundUnavailable'; readonly text: string }

export interface PlayableHarmonica {
    readonly holes: readonly HoleViewState[]
    readonly key: KeyViewState
    readonly style: PlayingStyleViewState
    readonly notesPerFinger: NotesPerFingerViewState
    readonly fingerMarks: FingerMarksViewState
    readonly demo: DemoViewState
    readonly toneShaping: ToneShapingViewState
}

export type LitHalf = 'top' | 'bottom'

export interface HoleViewState {
    readonly id: number
    readonly label: string
    readonly note: string
    readonly effect: string
    readonly lit: LitHalf | null
}

export interface KeyViewState {
    readonly label: string
    readonly position: number
    readonly highestPosition: number
}

export type PlayingStyleChoice =
    | 'severalFingersSeveralNotes'
    | 'severalFingersOneNote'
    | 'oneFingerSeveralNotes'
    | 'oneFingerNotesByPressure'

export const playingStyleChoices: readonly PlayingStyleChoice[] =
    ['severalFingersSeveralNotes', 'severalFingersOneNote', 'oneFingerSeveralNotes', 'oneFingerNotesByPressure']

export function isPlayingStyleChoice(value: string): value is PlayingStyleChoice {
    return playingStyleChoices.some(choice => choice === value)
}

export interface PlayingStyleChoiceViewState {
    readonly id: PlayingStyleChoice
    readonly name: string
}

export interface PlayingStyleViewState {
    readonly chosen: PlayingStyleChoice
    readonly label: string
    readonly choices: readonly PlayingStyleChoiceViewState[]
}

export interface NotesPerFingerViewState {
    readonly chosen: number
    readonly choices: readonly NoteCountViewState[]
    readonly isAvailable: boolean
}

export interface NoteCountViewState {
    readonly count: number
    readonly name: string
}

export interface FingerMarksViewState {
    readonly width: FingerWidthViewState
    readonly drawsOnlyTheDecidingFinger: boolean
}

export type FingerWidthViewState =
    | { readonly kind: 'resting' }
    | { readonly kind: 'holesWide'; readonly holes: number }
    | { readonly kind: 'pressed' }

export interface TuneViewState {
    readonly id: number
    readonly name: string
}

export interface DemoViewState {
    readonly label: string
    readonly tunes: readonly TuneViewState[]
    readonly isPlaying: boolean
}

export interface ToneShapingViewState {
    readonly overbendLabel: string
    readonly bendLabel: string
    readonly vibratoLabel: string
    readonly isOverbendAvailable: boolean
    readonly isBendAvailable: boolean
}
